src/provider/utils/async.ts

```diff
@@ -79,6 +79,48 @@ export function allowReject<T>(promise: Promise<T>): Promise<T> {
     return promise;
 }
 
+export async function raceUntilTrue<T>(items: T[], predicate: (v?: T) => boolean): Promise<T | null> {
+    let promisesCompleted = 0;
+    let errors = 0;
+    const valueFound = new DeferredPromise<T | null>();
+
+    const handleResolve = async (value?: T): Promise<T | void> => {
+        if (predicate(value)) {
+            valueFound.resolve(value);
+        }
+    };
+
+    const handleReject = (): void => {
+        errors++;
+    };
+
+    const completeCheck = async (value: T | void): Promise<T | null> => {
+        promisesCompleted++;
+        if (value) {
+            return value;
+        }
+        if (items.length === promisesCompleted) {
+            if (items.length === errors) {
+                throw new Error('All promises rejected');
+            }
+            valueFound.resolve(null);
+        }
+        await valueFound.promise;
```

Comment:
Is this needed?

Reply:
It needs to be awaited otherwise it will return and end the race.

Reply:
I wonder if this function should `Promise.all` rather than `Promise.race` then, then I think you *could* get rid of this line. And I thiiiink you could get rid of `racePromises.push(valueFound.promise);`.

But hold off changing until I've looked into reconciling this and develop!